### ISSUE TITLE
Intermittent Unit Test Failure - AutoRefreshingCacheMap.testMassRun()

### ISSUE THREAD

**Issue by warlockza** [promregator/promregator#127]
Intermittently when building against master, I get a fairly frequent failure of the this test
<pre>
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at org.cloudfoundry.promregator.cache.AutoRefreshingCacheMapTest.testMassRun(AutoRefreshingCacheMapTest.java:166)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
</pre>

Due to 
<pre>
Exception in thread "Thread-2" java.lang.IllegalThreadStateException
	at java.lang.Thread.start(Thread.java:708)
	at org.cloudfoundry.promregator.cache.AutoRefreshingCacheMap.ensureRefresherThreadIsRunning(AutoRefreshingCacheMap.java:92)
	at org.cloudfoundry.promregator.cache.AutoRefreshingCacheMap.put(AutoRefreshingCacheMap.java:161)
	at org.cloudfoundry.promregator.cache.AutoRefreshingCacheMapTest$MassOperationTestThread.run(AutoRefreshingCacheMapTest.java:116)
</pre>
Looking into the reason for failure, its because its trying to start a thread that is already started.

if I make the below method in AutoRefeshingCacheMap.java syncronised, then it reliably passes
```java
	private synchronized void ensureRefresherThreadIsRunning() {
		if (this.refresherThread == null) {
			this.refresherThread = new RefresherThread<K, V>(this);
			this.refresherThread.start();
		}
	}
```
I was considering using a DCL , but making it synchronized is cleaner 

Just a Side Note.. this only started failing recently some time after 0.6.1

**Comment by eaglerainbow** [comment]
Thanks for reporting this.

> Just a Side Note.. this only started failing recently some time after 0.6.1

Well, it's even a little more - I am being battling issues with this test for quite a while. I already was about to remove/disable it  due to this unreliablity.

**Comment by eaglerainbow** [comment]
I have included your approach with 7094a69
